### ISSUE TITLE
Tardigrade: Support ARM architectures [$5]

### ISSUE THREAD

**Issue by iSOcH** [duplicati/duplicati#4342]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: duplicati-2.0.5.111_canary_2020-09-26.spk
- **Operating system**: Linux 3.2.40 #25426 SMP Wed Jul 8 03:14:12 CST 2020 armv7l GNU/Linux synology_armadaxp_ds414
- **Backend**: Tardigrade / storj
- **Hardware**: Synology DS 414
-  **Mono**: Mono JIT compiler version 5.20.1.34 (tarball Mon Sep 28 18:52:24 UTC 2020)

## Description
Neither connection test nor backup to tardigrade appears to work on my NAS. Please note that I was not previously using duplicati, so I might also do something wrong.

## Steps to reproduce
1. Install mono on synology
2. Install duplicati using spk
3. Ran duplicati manually with command `mono /volume1/@appstore/Duplicati/Duplicati.Server.exe --webservice-interface=*` to be able to connect to Web UI
3. Prepare tardigrade bucket & api key
4. Create new backup job, specified
    - Storage type: Tardigrade
    - Auth: API key
    - Satellite: Satellite my account is on (europe-west)
    - API key
    - Encryption passphrase
    - Previously created bucket
    - Simple name for folder path
5. Click "Test connection"

- **Actual result**:
Error: Failed to connect: Method has no body
- **Expected result**:
Successful connection

The same error occurs also when trying to run the backup. Settings made for tardigrade storage appear not to matter.

## Debug log
```
System.BadImageFormatException: Method has no body
File name: 'uplink.NET'
  at uplink.SWIG.Config..ctor () [0x00000] in <ff23bbbb11844bf4b34766b871c2049c>:0
  at uplink.NET.Models.Config.ToSWIG () [0x00000] in <ff23bbbb11844bf4b34766b871c2049c>:0
  at uplink.NET.Models.Access.Init (uplink.NET.Models.Config config) [0x0003a] in <ff23bbbb11844bf4b34766b871c2049c>:0
  at uplink.NET.Models.Access..ctor (System.String satelliteAddress, System.String apiKey, System.String secret, uplink.NET.Models.Config config) [0x00006] in <ff23bbbb11844bf4b34766b871c2049c>:0
  at Duplicati.Library.Backend.Tardigrade.Tardigrade..ctor (System.String url, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x000bc] in <ad731a73693342f58ae6d6e9359b7d5e>:0
  at (wrapper managed-to-native) System.Reflection.MonoCMethod.InternalInvoke(System.Reflection.MonoCMethod,object,object[],System.Exception&)
  at System.Reflection.MonoCMethod.InternalInvoke (System.Object obj, System.Object[] parameters, System.Boolean wrapExceptions) [0x00005] in <4a89e81f3d9644e6a5d556824ae3eaf1>:0
--- End of stack trace from previous location where exception was thrown ---

  at Duplicati.Library.DynamicLoader.BackendLoader+BackendLoaderSub.GetBackend (System.String url, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x0017e] in <d8475073106c4c8181064ee275fcff6d>:0
  at Duplicati.Library.DynamicLoader.BackendLoader.GetBackend (System.String url, System.Collections.Generic.Dictionary`2[TKey,TValue] options) [0x00000] in <d8475073106c4c8181064ee275fcff6d>:0
  at Duplicati.Server.WebServer.RESTMethods.RemoteOperation.TestConnection (System.String url, Duplicati.Server.WebServer.RESTMethods.RequestInfo info) [0x000b0] in <d348e18e6cb747ab98d1065177fd004f>:0
  at Duplicati.Server.WebServer.RESTMethods.RemoteOperation.POST (System.String key, Duplicati.Server.WebServer.RESTMethods.RequestInfo info) [0x00094] in <d348e18e6cb747ab98d1065177fd004f>:0
  at Duplicati.Server.WebServer.RESTHandler.DoProcess (Duplicati.Server.WebServer.RESTMethods.RequestInfo info, System.String method, System.String module, System.String key) [0x00289] in <d348e18e6cb747ab98d1065177fd004f>:0
```

**Comment by warwickmm** [comment]
/cc @TopperDEL 

Could this be due to some architecture incompatibility?

**Comment by TopperDEL** [comment]
It seems that Duplicati/.Net tries to load the x64-DLL from x86 or vice versa.
@iSOcH Do you know what plattform you are on?

I'm not quite sure how to solve this, though. The DLL-deployment might be wrong or we need to add a switch like we have on Windows to use the right binary. But I have no way to test this on my end.

**Comment by iSOcH** [comment]
From the info above: Linux 3.2.40 #25426 SMP Wed Jul 8 03:14:12 CST 2020 *armv7l* GNU/Linux synology_*armadaxp*_ds414

And here is my /proc/cpuinfo:
```
Processor       : Marvell PJ4Bv7 Processor rev 2 (v7l)
processor       : 0
BogoMIPS        : 1332.01

processor       : 1
BogoMIPS        : 1332.01

Features        : swp half thumb fastmult vfp edsp vfpv3 tls
CPU implementer : 0x56
CPU architecture: 7
CPU variant     : 0x2
CPU part        : 0x584
CPU revision    : 2

Hardware        : Marvell Armada XP Development Board
Revision        : 0000
Serial          : 0000000000000000
```

Can I help with other information?

**Comment by TopperDEL** [comment]
Ah, sorry - I oversaw the "arm" - so you have an architecture that currently is not available in the used uplink.NET-library. I would need to compile an arm-version of it and deploye it here.
I'm sorry, this is something I cannot tackle currently. I need to postpone it. It should be possible, though.

**Comment by toconspiracy** [comment]
Same Issue here, want to use Duplicati with Tardigrade on ARM (Using the Official Duplicati Dockerimage). Any ETA when you can tackle that Issue down? (or something i can help/test?)

**Comment by warwickmm** [comment]
Thanks @toconspiracy.  I've edited the issue title to better describe what's needed.

**Comment by TopperDEL** [comment]
@toconspiracy I would like to do this all sooner than later. But Storj is currently reworking their uplink, where my wrapper is based on. And the told me that I should wait with any further adjustments for the next release as it will break some things anyway. So if you ask me I would say this won't happen before ~february.

But I have it on my list and will inform you on this issue once this gets further or is done!

**Comment by TopperDEL** [comment]
To get this issue further I unfortunately need some adjustments to Go - see [this issue here](https://github.com/golang/go/issues/43800). Once this is done I can add that binary to uplink.Net to ultimately support the Tardigrade-Backend of Duplicati for ARM.

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/storj-uplink-dll/11733/3

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/backup-on-tardigrade/12295/3

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/battle-plan-for-migrating-to-net8/17482/22

**Comment by TopperDEL** [comment]
Finally there should be an ARM-working version of Duplicati for Storj (formerly known as Tardigrade) backend on the horizon! Duplicati needs to update to uplink.NET 2.13.* and Linux should be supported on x64, arm and arm64.

**Comment by kenkendk** [comment]
The Arm64 packages are now in the canary builds, so closing this issue
